### PR TITLE
Use WebSocket wrapper for all managers and clients

### PR DESCRIPTION
**Pull request by Nerixyz**
# Description

As mentioned in https://github.com/Chatterino/chatterino2/issues/3452#issuecomment-1635860032, `websocketpp` causes slow compilation times. This PR attempts to combat this by encapsulating all `websocketpp` uses in a single translation unit (`providers/ws/Client.cpp`). Both PubSub and Liveupdates use this under the hood now.

### CODE REVIEW

File: src/providers/ws/Client.cpp [hunk 1]
```diff
@@ -0,0 +1,298 @@
...
+void ClientPrivate::onMessage(WebsocketppHandle &&hdl,
+                              const WebsocketppMessagePtr &msg)
+{
+    Q_Q(Client);
+
+    // TODO(Qt6): this should be a QByteArrayView
+    q->onTextMessage(
+        std::move(hdl),
+        QLatin1String(msg->get_payload().c_str(),
+                      static_cast<qsizetype>(msg->get_payload().size())));
```
[the remark this answers is not in the document]
Comment (Nerixyz) [reply src/providers/ws/Client.cpp:131]: qt5 thing

File: src/providers/ws/Client.cpp [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+    void onConnectionOpen(WebsocketppHandle &&hdl);
+    void onConnectionClose(WebsocketppHandle &&hdl);
+    void onConnectionFail(WebsocketppHandle &&hdl);
+
+    void runThread();
+
+    Client *q_ptr;
+    Q_DECLARE_PUBLIC(Client)
+
+    std::shared_ptr<boost::asio::io_service::work> work_{nullptr};
```
Comment (Nerixyz) [inline src/providers/ws/Client.cpp:96]: This is deprecated. I should change it.

**Review by pajlada — CHANGES_REQUESTED** [review]
Same feedback on the pimpl stuff as the nativemessaging refactor, otherwise the idea is fine. will review more once the Q_Q's are gone

File: src/providers/ws/Client.cpp [hunk 3]
```diff
@@ -0,0 +1,319 @@
...
+
+Client::~Client()
+{
+    this->forceStop();
+};
+
+void Client::start()
+{
+    auto *d = this->private_.get();
+
```
Comment (pajlada) [inline src/providers/ws/Client.cpp:188]: Assert that asioThread is not set

File: src/providers/ws/Client.cpp [hunk 4]
```diff
@@ -0,0 +1,319 @@
...
+    d->asioThread = std::make_unique<std::thread>([d] {
+        d->runThread();
+    });
+}
+
+void Client::stop()
+{
+    auto *d = this->private_.get();
+
```
Comment (pajlada) [inline src/providers/ws/Client.cpp:197]: Assert that asioThread is set

File: src/providers/ws/Client.cpp [hunk 5]
```diff
@@ -0,0 +1,319 @@
...
+    if (d->asioThread->joinable())
+    {
+        d->asioThread->join();
+    }
+}
+
+void Client::forceStop()
+{
+    auto *d = this->private_.get();
+
```
Comment (pajlada) [inline src/providers/ws/Client.cpp:212]: Assert that asioThread is set

File: src/providers/ws/Client.cpp [hunk 6]
```diff
@@ -0,0 +1,319 @@
...
+    WebsocketppErrorCode ec;
+    d->websocketClient.send(conn.hdl_, str, len,
+                            websocketpp::frame::opcode::text, ec);
+
+    if (ec)
+    {
+        qCDebug(chatterinoLiveupdates)
```
Comment (pajlada) [inline src/providers/ws/Client.cpp:254]: Fix the logigng categories in the file

**Review by pajlada — CHANGES_REQUESTED** [review]
Small nitpicks, overall looks good :+1:

**Comment by Nerixyz** [comment]
Gonna close this, since we're moving to eventsub which uses beast. We might want to have a wrapper around that (to reduce compilation time) in the future.

**Pull request closed without merging**